### PR TITLE
egmde.remote: a handy launch script for using egmde over VNC

### PR DESCRIPTION
**Pull request by AlanGriffiths**
This sets up ssh tunnelling and a remote VNC server, once the  server is running it can be connected to locally with any VNC  client (e.g. `gvncviewer localhost`)

### CODE REVIEW

**Comment by AlanGriffiths** [comment]
This is somewhat of a "request for comments", it can likely be improved in a number of ways. But is also useful "as is".

File: glue/egmde.remote [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  echo "  This sets up ssh tunnelling and a remote VNC server, once the"
+  echo "  server is running it can be connected to locally with any VNC"
+  echo "  client (e.g. 'gvncviewer localhost')"
+  echo "Usage: $(basename $0) [ssh-login]"
+  echo "[ssh-login] login for ssh: hostname, user@hostname or -p port user@ipaddress"
+  echo "The desktop size can be set with EGMDE_REMOTE_OUTPUT=WxH $(basename $0) [login]"
+  exit
+fi
+
+ssh -T -L 5900:localhost:5900 $* -- bash -el <<EOT
```
Comment (Saviq) [inline glue/egmde.remote]: Shouldn't we pass the args as is?

```suggestion
ssh -T -L 5900:localhost:5900 "$@" -- bash -el <<EOT
```

**Review by Saviq — CHANGES_REQUESTED** [review]
Any way you can think of dealing with port 5900 being busy locally and/or remotely?

File: glue/egmde.remote (hunk 1 above)
Comment (AlanGriffiths) [reply glue/egmde.remote]: I think that failed with `-p 10022 alan-griffiths@localhost`. (Will check later)

**Comment by AlanGriffiths** [comment]
> Any way you can think of dealing with port 5900 being busy locally and/or remotely?

That gets more complicated as the the ssh tunnel needs to reflect the port. Which means trying to co-ordinate two ssh sessions. Possible, but significantly more work. (Maybe just find a way to detect and error out with an additional option to specify a port?)

Thoughts?

File: glue/egmde.remote (hunk 1 above)
Comment (AlanGriffiths) [reply glue/egmde.remote]: No, you are right! (I know, you knew that.)

**Comment by Saviq** [comment]
> (Maybe just find a way to detect and error out with an additional option to specify a port?)

That would be plenty. Since the same SSH process is doing the tunnel and executing it on the remote side, that sounds tricky, though… But maybe if you split it up into two processes?

**Comment by AlanGriffiths** [comment]
> > (Maybe just find a way to detect and error out with an additional option to specify a port?)
> 
> That would be plenty.

Done

**Comment by Saviq** [comment]
When I ran this to connect to a host that already had an X session, that's what I got when connecting to the requested port… Any way for it to notice to use a different DISPLAY?

**Comment by AlanGriffiths** [comment]
On Mon, 28 Mar 2022 06:26:28 -0700, Michał Sawicz wrote:
> When I ran this to connect to a host that already had an X 
> session, that's what I got when connecting to the requested 
> port… Any way for it to notice to use a different DISPLAY?
> 

Yes, if you use, say, 5913 then that uses DISPLAY=13.

-- 
Sent using Dekko from my Ubuntu device

**Comment by Saviq** [comment]
> Yes, if you use, say, 5913 then that uses DISPLAY=13.

Hmm that's not my experience.

I ran this:
```plain
$ EGMDE_REMOTE_RFBPORT=5901 EGMDE_REMOTE_LOCALPORT=5902 egmde.remote localhost
```

And saw my `DISPLAY=:0` X session:
```plain
# ...
The VNC desktop is:      michal-laptop:1                                                                         
# ...
29/03/2022 08:55:44 Using X display :0
```

**Comment by AlanGriffiths** [comment]
> Hmm that's not my experience.

I'm back at a computer, let me try to reproduce

[edit]

So when I try this I get:
```
$ EGMDE_REMOTE_RFBPORT=5901 EGMDE_REMOTE_LOCALPORT=5902 egmde.remote localhost
...
The VNC desktop is:      octopull-PROX15-20:1
...
29/03/2022 09:47:41 wait_for_client: running: env X11VNC_SKIP_DISPLAY=''  /bin/sh /tmp/x11vnc-find_display.agCGHy
/tmp/x11vnc-find_display.agCGHy: 1: netstat: not found
29/03/2022 09:47:41 wait_for_client: find display cmd failed.
29/03/2022 09:47:41 wait_for_client: FINDCREATEDISPLAY cmd: /bin/sh /tmp/x11vnc-find_display.agCGHy Xvfb
trying N=20 ...
redir_daemon=
/usr/bin/xinit egmde.login -- /usr/bin/Xvfb :20 -screen 0 1280x1028x24 -cc 4 -nolisten tcp -auth /tmp/.xas4339222911.KfUatU 

/usr/bin/xinit: XFree86_VT property unexpectedly has 0 items instead of 1
29/03/2022 09:47:42 Using X display :20

```

Which "works" but is not exactly what I expected, but not what you're seeing either.

image

**Review by Saviq — APPROVED** [review]
A'ight, good enough!

**Comment by Saviq** [comment]
bors merge

**Pull request merged**